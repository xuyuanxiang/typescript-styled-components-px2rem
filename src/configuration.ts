export interface IConfiguration {
  rootValue: number;
  unitPrecision: number;
  minPixelValue: number;
  multiplier: number;
  tags: ReadonlyArray<string>;
  propWhiteList: ReadonlyArray<string>;
  propBlackList: ReadonlyArray<string>;
  exclude: boolean;
  selectorBlackList: ReadonlyArray<string>;
  ignoreIdentifier: boolean;
  replace: boolean;
  mediaQuery: boolean;
  transformRuntime: boolean;
}

class ConfigurationManager {
  private static readonly defaultConfiguration: IConfiguration = {
    rootValue: 100,
    unitPrecision: 5,
    minPixelValue: 2,
    multiplier: 1,
    tags: ['styled', 'css', 'createGlobalStyle', 'keyframes'],
    propWhiteList: [],
    propBlackList: [],
    exclude: false,
    selectorBlackList: [],
    ignoreIdentifier: false,
    replace: true,
    mediaQuery: false,
    transformRuntime: false,
  };
  private _config: IConfiguration = ConfigurationManager.defaultConfiguration;

  public get config(): IConfiguration {
    return this._config;
  }

  public updateConfig(config?: Partial<IConfiguration>): void {
    if (config) {
      this._config = Object.assign({}, ConfigurationManager.defaultConfiguration, config);
    }
  }
}

export default new ConfigurationManager();
