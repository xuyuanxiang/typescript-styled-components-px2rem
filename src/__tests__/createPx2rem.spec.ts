import * as ts from 'typescript';
import { runInNewContext } from 'vm';
import createPx2rem from '../createPx2rem';
import configuration, { IConfiguration } from '../configuration';

function px2rem(
  value: string,
  {
    rootValue = configuration.config.rootValue,
    unitPrecision = configuration.config.unitPrecision,
    multiplier = configuration.config.multiplier,
    minPixelValue = configuration.config.minPixelValue,
  }: Partial<IConfiguration> = configuration.config,
): string {
  const sandbox = { result: '' };
  const transformed = ts.transpileModule('', {
    compilerOptions: {
      skipLibCheck: true,
      target: ts.ScriptTarget.ES2015,
      module: ts.ModuleKind.ES2015,
      strict: true,
      esModuleInterop: true,
    },
    transformers: {
      before: [
        context => node => {
          node = ts.updateSourceFileNode(node, [
            createPx2rem(ts.createIdentifier('px2rem'), { rootValue, unitPrecision, multiplier, minPixelValue }),
          ]);
          return node;
        },
      ],
    },
  });
  runInNewContext(`${transformed.outputText} result = px2rem(${value});`, sandbox);
  return sandbox.result;
}

it('should match snapshot', function() {
  const transformed = ts.transpileModule('', {
    compilerOptions: {
      skipLibCheck: true,
      target: ts.ScriptTarget.ES2015,
      module: ts.ModuleKind.ES2015,
      strict: true,
      esModuleInterop: true,
    },
    transformers: {
      before: [
        context => node => {
          node = ts.updateSourceFileNode(node, [createPx2rem(ts.createIdentifier('px2rem'), configuration.config)]);
          return node;
        },
      ],
    },
  });
  expect(transformed.outputText).toMatchSnapshot();
});

it('should transform String', function() {
  expect(px2rem("'-100px'")).toBe('-1rem');
  expect(px2rem("'32px'")).toBe('0.32rem');
  expect(px2rem("'11.3333333px'")).toBe('0.11333rem');
  expect(px2rem("'-11.3333333px'")).toBe('-0.11333rem');
});

it('should transform Number', function() {
  expect(px2rem('16')).toBe('0.16rem');
  expect(px2rem('16.3333333')).toBe('0.16333rem');
  expect(px2rem('-16')).toBe('-0.16rem');
  expect(px2rem('-16.3333333')).toBe('-0.16333rem');
});

it('should transform Function', function() {
  expect(px2rem('function() {return 20;}')).toBe('0.2rem');
  expect(px2rem('function(props) { return props.width / 2; }, { width: 100 }')).toBe('0.5rem');
});

it('should ignore value less than "minPixelValue" option', function() {
  const options = { minPixelValue: 2 };
  expect(px2rem("'1px'", options)).toBe('1px');
  expect(px2rem('1', options)).toBe('1px');
  expect(px2rem("'1.3333333px'", options)).toBe('1.3333333px');
  expect(px2rem('1.3333333', options)).toBe('1.3333333px');
  expect(px2rem("'-1px'", options)).toBe('-1px');
  expect(px2rem('-1', options)).toBe('-1px');
  expect(px2rem("'-1.3333333px'", options)).toBe('-1.3333333px');
  expect(px2rem('-1.3333333', options)).toBe('-1.3333333px');
  expect(px2rem('function() {return 1;}', options)).toBe('1px');
  expect(px2rem('function() {return "1px";}', options)).toBe('1px');
  expect(px2rem('function(props) { return props.width / 2; }, { width: 2 }', options)).toBe('1px');
  expect(px2rem('function(props) { return props.width / 2 + "px"; }, { width: 2 }', options)).toBe('1px');
});

it('should multiply "multiplier" option', function() {
  const options = { minPixelValue: 2, multiplier: 2 };
  expect(px2rem("'1px'", options)).toBe('1px');
  expect(px2rem('1', options)).toBe('1px');
  expect(px2rem("'1.3333333px'", options)).toBe('1.3333333px');
  expect(px2rem('1.3333333', options)).toBe('1.3333333px');
  expect(px2rem("'-1px'", options)).toBe('-1px');
  expect(px2rem('-1', options)).toBe('-1px');
  expect(px2rem("'-1.3333333px'", options)).toBe('-1.3333333px');
  expect(px2rem('-1.3333333', options)).toBe('-1.3333333px');
  expect(px2rem('function() {return 1;}', options)).toBe('1px');
  expect(px2rem('function() {return "1px";}', options)).toBe('1px');
  expect(px2rem('function(props) { return props.width / 2; }, { width: 2 }', options)).toBe('1px');
  expect(px2rem('function(props) { return props.width / 2 + "px"; }, { width: 2 }', options)).toBe('1px');
  expect(px2rem("'2px'", options)).toBe('0.04rem');
  expect(px2rem('2', options)).toBe('0.04rem');
  expect(px2rem("'2.3333333px'", options)).toBe('0.04667rem');
  expect(px2rem("'-2.3333333px'", options)).toBe('-0.04667rem');
  expect(px2rem('2.3333333', options)).toBe('0.04667rem');
  expect(px2rem('-2.3333333', options)).toBe('-0.04667rem');
  expect(px2rem("'-2px'", options)).toBe('-0.04rem');
  expect(px2rem('-2', options)).toBe('-0.04rem');
  expect(px2rem('function() {return 2;}', options)).toBe('0.04rem');
  expect(px2rem('function() {return "2px";}', options)).toBe('0.04rem');
  expect(px2rem('function(props) { return props.width / 2; }, { width: 4 }', options)).toBe('0.04rem');
  expect(px2rem('function(props) { return props.width / 2 + "px"; }, { width: 4 }', options)).toBe('0.04rem');
});
