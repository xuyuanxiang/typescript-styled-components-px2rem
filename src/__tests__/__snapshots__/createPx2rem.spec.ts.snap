// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
"\\"use strict\\";
function px2rem(input, ...args) { if (typeof input === \\"function\\")
    return px2rem(input(...args), ...args); var value = parseFloat(input); var pixels = Number.isNaN(value) ? 0 : value; if (Math.abs(pixels) < 0)
    return \`\${pixels}px\`; var multiplier = Math.pow(10, 5 + 1); var wholeNumber = Math.floor((pixels * 1) / 100 * multiplier); return \`\${(Math.round(wholeNumber / 10) * 10) / multiplier}rem\`; }
"
`;
